Migrate WPK directory to wazuh/wazuh repository

jotacarma90 opened the pull request:
|Related issue|
|---|
|https://github.com/wazuh/wazuh/issues/22321|



## Description

This PR is to add the WPK package generation directory of the migration from wazuh-packages to wazuh/wazuh.
We are also adding some workflows from GitHub actions to generate them.

### Resume changes

After all the migration of the WPK generation files. Some changes have been made and new files created:
- `generate_wpk_package.sh` script:
  - Added AWS secrets environment variables to docker run command.
  - Added var `MOUNT_KEYDIR_FLAG` to avoid errors when keys dir is not set in parameters.
  - Added docker tag name when building Docker image.
  - Added `--dont-build-docker` parameter to avoid building docker image.
  - Added warnings in parameters to warn about unsupported architectures.

- Workflow to Docker image upload.
<details>
<summary>Inputs:</summary>

```yml
      tag:
        description: 'Tag name of the Docker image to be uploaded (Use developer to set branch name. Use auto to set src/VERSION content)'
        required: false
        default: auto
```

</details>

- Workflow to build WPK packages (Permission request to access WPK secrets: https://github.com/wazuh/wazuh-jenkins/issues/6344):
<details>
<summary>Inputs:</summary>

```yml
      tag:
        description: |
          Tag name of the Docker image to be downloaded.
          Use 'developer' to set branch name as tag.
          Use 'auto' to set branch version as tag.
          Default is 'auto'.
        required: false
        default: 'auto'
      linux_branch:
        description: |
          Branch name for compiling the Linux WPK.
          If empty, it will not be generated.
        required: false
      windows_package:
        description: |
          Windows WPK name in S3 or link to download.
          If empty, it will not be generated.
        required: false
      macos_package:
        description: |
          MacOS WPK name in S3 or link to download.
          If empty, it will not be generated.
        required: false
      revision:
        description: |
          Revision used to naming WPK package.
          Default is '0'.
        required: false
        default: '0'
      release_name_format:
        description: |
          True if WPK name should have release format.
          False if WPK name should have developer format.
          Default is 'false'.
        required: false
        type: boolean
        default: false
```

</details>

jotacarma90 commented:
Issue blocked by https://github.com/wazuh/internal-devel-requests/issues/964
Extending ETA waiting for AWS secrets permissions.

mjcr99 commented on .github/workflows/packages-build-wpk.yml:
```diff
@@ -0,0 +1,98 @@
+  workflow_dispatch:
+    inputs:
+      tag:
+        description: 'Tag name of the Docker image to be downloaded (Use developer to set branch name. Use auto to set src/VERSION content)'
```
The way this input is set, allows values different from `developer` or `auto`, should we restrict it?

jotacarma90 replied:
I think yes, users could want add a tag name referring some custom names.

mjcr99 commented on .github/workflows/packages-build-wpk.yml:
```diff
@@ -0,0 +1,98 @@
+        required: false
+        default: 1
+
+  workflow_call:
```
If this workflow is not going to be called from another workflow, we could remove this `workflow_call` tag.

jotacarma90 replied:
done

MarcelKemp commented on .github/workflows/packages-build-wpk.yml:
```diff
@@ -0,0 +1,98 @@
+          COMMIT_HASH="$(curl -s https://api.github.com/repos/wazuh/wazuh/commits/${{ github.ref_name }} | grep '"sha"' | head -n 1| cut -d '"' -f 4 | cut -c 1-7)"
+          echo "WPK_NAME=wazuh-wpk_${VERSION}-${{ github.event.inputs.revision }}_${{ matrix.os }}_${{ matrix.arch }}_${COMMIT_HASH}.wpk" >> $GITHUB_ENV
+          if [ "${{ matrix.os }}" == "linux" ]; then echo "CONTAINER_NAME=linux_wpk_builder_x86_64" >> $GITHUB_ENV; else echo "CONTAINER_NAME=common_wpk_builder" >> $GITHUB_ENV; fi
+
```
Could we add a **_step_** to verify that a package similar to the one we want to build does not already exist in the **S3** repository? 
In case it does exist, indicate link and do not build the package again.

jotacarma90 replied:
Done

MarcelKemp commented on .github/workflows/packages-build-wpk.yml:
```diff
@@ -0,0 +1,98 @@
+        description: 'Revision used to naming WPK package'
+        required: false
+        default: 1
+
```
Let's add a `release` parameter, so that when set to `true`, it uses the current nomenclature for the release.

lchico replied:
Can we use this parameter to define:

"0" for development
"1" for release
Issue number for custom ID

jotacarma90 replied:
Done, added new parameter release_name_format to better clarify to the users.

MarcelKemp commented on packages/wpk/README.md:
Let's update the _README.md_ with the information currently available, and add extra information that may be useful.

mjcr99 commented on .github/workflows/packages-build-wpk.yml:
```diff
@@ -0,0 +1,98 @@
+        if: matrix.os == 'linux' && github.event.inputs.linux_branch != ''
+        working-directory: packages/wpk
+        run: |
+          bash generate_wpk_package.sh -t linux --dont-build-docker --tag ${{ env.TAG }} -b ${{ github.event.inputs.linux_branch }} -d /tmp --aws-wpk-key WPK_PRIVATE_KEY --aws-wpk-cert WPK_CERT -o ${{ env.WPK_NAME }}
```
By default the `generate_wpk_package.sh` script is run with `JOBS=4`. Should we add a `-j $(nproc)` call argument to modify this parameter according to the wpk building machine?

lchico commented on .github/workflows/packages-build-wpk.yml:
```diff
@@ -0,0 +1,98 @@
+  WPK-generation:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
```
Do we always need to generate all the wpk packages? It would be helpful to be able to generate only one or two specific wpk packages.

jotacarma90 replied:
this functionality is allowed by leaving the package parameter of each system empty respectively

lchico commented on .github/workflows/packages-upload-wpk-images.yml:
```diff
@@ -0,0 +1,44 @@
+  Upload-WPK-images:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
```
Similar to the packages, do we need to deploy all the containers? It would be beneficial to be able to select specific containers for deployment.

jotacarma90 replied:
done

lchico commented on packages/wpk/generate_wpk_package.sh:
```diff
@@ -0,0 +1,381 @@
+                shift 2
+            fi
+            ;;
+        "-c"|"--checksum")
```
The current logic assumes that if there's an argument after -c or --checksum, it must be the checksum directory. However, if another option follows with its own argument (like -s /tmp), the script will attempt to use /tmp as the checksum directory, leading to an error.

jotacarma90 replied:
added new control to this parameters, but we want to maintain checksum dir parameter i think

lchico reviewed (changes requested):
Great job @jotacarma90, I left some comments.

jotacarma90 commented:
# Resume changes

After all the migration of the WPK generation files. Some changes have been made and new files created:
- `generate_wpk_package.sh` script:
  - Added AWS secrets environment variables to docker run command.
  - Added var `MOUNT_KEYDIR_FLAG` to avoid errors when keys dir is not set in parameters.
  - Added docker tag name when building Docker image.
  - Added `--dont-build-docker` parameter to avoid building docker image.
  - Added warnings in parameters to warn about unsupported architectures.

- Workflow to Docker image upload.
  - inputs:
```
      tag:
        description: 'Tag name of the Docker image to be uploaded (Use developer to set branch name. Use auto to set src/VERSION content)'
        required: false
        default: auto
```
- Workflow to build WPK packages (Permission request to access WPK secrets: https://github.com/wazuh/wazuh-jenkins/issues/6344):
  - inputs:
```
      tag:
        description: |
          Tag name of the Docker image to be downloaded.
          Use 'developer' to set branch name as tag.
          Use 'auto' to set branch version as tag.
          Default is 'auto'.
        required: false
        default: 'auto'
      linux_branch:
        description: |
          Branch name for compiling the Linux WPK.
          If empty, it will not be generated.
        required: false
      windows_package:
        description: |
          Windows WPK name in S3 or link to download.
          If empty, it will not be generated.
        required: false
      macos_package:
        description: |
          MacOS WPK name in S3 or link to download.
          If empty, it will not be generated.
        required: false
      revision:
        description: |
          Revision used to naming WPK package.
          Default is '0'.
        required: false
        default: '0'
      release_name_format:
        description: |
          True if WPK name should have release format.
          False if WPK name should have developer format.
          Default is 'false'.
        required: false
        type: boolean
        default: false
```

jotacarma90 commented:
# Update 01/04/2024

I have modified the names of some input variables.
Updated some flows
Removed the package existence check. It will always be overwritten.
Successful testing of the workflow with different parameters:
- https://github.com/wazuh/wazuh/actions/runs/8507481782/job/23299397688
- https://github.com/wazuh/wazuh/actions/runs/8507648406/job/23299854467

MarcelKemp reviewed (approved):
LGTM

mjcr99 reviewed (approved):
LGTM
